storage/tests/test_isolation.rs

```diff
@@ -123,3 +127,488 @@ fn g0_update_conflicts_fail() {
+            reader_after_2.close_resources();
+        }
+        Err(_) => assert!(false)
+    }
+
+    // Continue
+    snapshot_1.put_val(key_2.to_owned(), ByteArray::copy(value_21.bytes()));
+    let result_1 = snapshot_1.commit();
+    match result_1 {
+        Ok(()) => {
+            let reader_after_1 = storage.open_snapshot_read();
+            assert_eq!(reader_after_1.get::<BUFFER_VALUE_INLINE>(StorageKeyReference::from(&key_1)).unwrap().bytes(), value_11.bytes());
+            assert_eq!(reader_after_1.get::<BUFFER_VALUE_INLINE>(StorageKeyReference::from(&key_2)).unwrap().bytes(), value_21.bytes());
+            // reader_after_1.close();
+        }
+        Err(_) => ()
+    }
+}
+
+#[test]
+fn g1a_aborted_writes() {
+    // This requires a wr edge which do not exist under snapshots
+    // With snapshots, concurrent transactions cannot see each others writes.
+    init_logging();
+    let key_1 = StorageKeyArray::new(KEYSPACE_ID, ByteArray::copy(&KEY_1));
+    let value_1_0 = ByteArray::inline([10], 1);
+    let value_1_1 = ByteArray::inline([11], 1);
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+
+    let snapshot_setup = storage.open_snapshot_write();
+    snapshot_setup.put_val(key_1.to_owned(), ByteArray::copy(value_1_0.bytes()));
+    let result_snapshot = snapshot_setup.commit();
+    assert!(result_snapshot.is_ok());
+
+    let snapshot_1 = storage.open_snapshot_write();
+    let snapshot_2 = storage.open_snapshot_write();
+    snapshot_1.put_val(key_1.to_owned(), ByteArray::copy(value_1_1.bytes()));
+    let read_2 = snapshot_2.get::<BUFFER_VALUE_INLINE>(StorageKeyReference::from(&key_1));
+    snapshot_1.close_resources();
+    assert_eq!(value_1_0.bytes(), read_2.unwrap().bytes());
+}
+
+#[test]
+fn g1b_intermediate_read() {
+    // This requires a wr edge which do not exist under snapshots
+    // Since all writes happen together at commit time, either the initial or the final value is read.
+    init_logging();
+    let key_1 = StorageKeyArray::new(KEYSPACE_ID, ByteArray::copy(&KEY_1));
+    let value_1_0 = ByteArray::inline([10], 1);
+    let value_1_1i = ByteArray::inline([111], 1);
+    let value_1_1f = ByteArray::inline([112], 1);
+    let storage_path = create_tmp_dir();
+    let storage = setup_storage(&storage_path);
+
+    let snapshot_setup = storage.open_snapshot_write();
+    snapshot_setup.put_val(key_1.to_owned(), ByteArray::copy(value_1_0.bytes()));
+    let result_snapshot = snapshot_setup.commit();
+    assert!(result_snapshot.is_ok());
+
+    let snapshot_1 = storage.open_snapshot_write();
+    let snapshot_2 = storage.open_snapshot_write();
+    snapshot_1.put_val(key_1.to_owned(), ByteArray::copy(value_1_1i.bytes()));
+    let read_2 = snapshot_2.get::<BUFFER_VALUE_INLINE>(StorageKeyReference::from(&key_1));
+    snapshot_1.put_val(key_1.to_owned(), ByteArray::copy(value_1_1f.bytes()));
```

Comment:
This fails because overwriting a key that was written in the same transaction is not yet implemented.